### PR TITLE
Make variable definition of 'foreach' loop final

### LINKED ISSUE #18169: [Spec Deviation] Variables created in foreach typed-binding-pattern should be final
**Description:**
The spec says

```
foreach-stmt :=
   foreach typed-binding-pattern in action-or-expr block-stmt
```

A foreach statement iterates over a sequence, executing a block statement once for each member of the sequence.

The scope of any variables created in typed-binding-pattern is block-stmt. These variables are implicitly final."

However, the implementation allows updating such variables.

**Steps to reproduce:**
```ballerina
import ballerina/io;

public function main() {
    int[] arr = [1, 2];

    foreach int i in arr {
        io:println(i);
        i = 10;
        io:println(i);
    }
}
```

```cmd
$ ballerina run test.bal 
1
10
2
10
```

**Affected Versions:**
```cmd
$ ballerina -v
Ballerina 1.0.0-alpha3
Language specification 2019R2
Ballerina tool 1.0.0
```

### PR DESCRIPTION
**Pull request by gimantha**
## Purpose

Fixes #18169

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [x] Added necessary documentation  
  - [x] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: langlib/langlib-test/src/test/resources/test-src/statements/foreach/foreach-negative.bal [hunk 1]
```diff
@@ -86,6 +119,72 @@ function test10(){
...
+    foreach var {id, name, salary} in data {
+        id = 2;
+        name = "John";
+        salary = 250.5;
+    }
+}
+
+function test14() returns string {
+    output = "";
+    json subjects =  <json>jdata.subjects;
```
Comment (Shan1024) [inline langlib/langlib-test/src/test/resources/test-src/statements/foreach/foreach-negative.bal]:
```suggestion
    json subjects = <json>jdata.subjects;
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 2]
```diff
@@ -867,6 +879,21 @@ private void recursivelyDefineVariables(BLangVariable variable, SymbolEnv blockE
         }
     }
 
+    private void recursivelySetFinalFlag(BLangVariable variable) {
+        switch (variable.getKind()) {
```
Comment (Shan1024) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Don't we need to check for `ERROR_VARIABLE` here? Is it final by default?

I think it is better to add a default clause as well which throws an error to make sure that no unexpected value gets ignored here. WDYT?

Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: @Shan1024 I thought i come under generic variable, as we are dealing with variables when restructured. Let me add a test case for that and verify

Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: you are right. it fails for error variable. Fixed it and added a test

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 3]
```diff
@@ -902,6 +914,30 @@ private void recursivelyDefineVariables(BLangVariable variable, SymbolEnv blockE
...
+            case VARIABLE:
+                variable.symbol.flags |= Flags.FINAL;
+                break;
+            case TUPLE_VARIABLE:
+                ((BLangTupleVariable) variable).memberVariables.forEach(this::recursivelySetFinalFlag);
+                break;
+            case RECORD_VARIABLE:
+                ((BLangRecordVariable) variable).variableList.forEach(value ->
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Don't we have to check rest param here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 4]
```diff
@@ -902,6 +914,30 @@ private void recursivelyDefineVariables(BLangVariable variable, SymbolEnv blockE
         }
     }
 
+    private void recursivelySetFinalFlag(BLangVariable variable) {
+        switch (variable.getKind()) {
+            case VARIABLE:
+                variable.symbol.flags |= Flags.FINAL;
+                break;
+            case TUPLE_VARIABLE:
+                ((BLangTupleVariable) variable).memberVariables.forEach(this::recursivelySetFinalFlag);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Shouldn't we check restVariable too?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 3 above)
Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 4 above)
Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: fixed

**Pull request merged**